api/client_test.go

```diff
@@ -0,0 +1,513 @@
+
+	t.Run("missing ltConfig", func(t *testing.T) {
+		agent, err := client.CreateAgent("someid", nil, nil)
+		require.EqualError(t, err, "agent: ltConfig should not be nil")
+		require.Nil(t, agent)
+	})
+
+	t.Run("missing ucConfig", func(t *testing.T) {
+		agent, err := client.CreateAgent("someid", &loadtest.Config{}, nil)
+		require.EqualError(t, err, "agent: ucConfig should not be nil")
+		require.Nil(t, agent)
+	})
+
+	t.Run("missing uc type", func(t *testing.T) {
+		agent, err := client.CreateAgent("someid", &loadtest.Config{}, &simulcontroller.Config{})
+		require.EqualError(t, err, "agent: UserController type is not set")
+		require.Nil(t, agent)
+	})
+
+	t.Run("invalid ucConfig type", func(t *testing.T) {
+		var ltConfig loadtest.Config
+		ltConfig.UserControllerConfiguration.Type = "simulative"
+		agent, err := client.CreateAgent("someid", &ltConfig, "invalid")
+		require.EqualError(t, err, "agent: ucConfig has the wrong type")
+		require.Nil(t, agent)
+	})
+
+	t.Run("invalid configs", func(t *testing.T) {
+		var ltConfig loadtest.Config
+		ltConfig.UserControllerConfiguration.Type = "simulative"
+		agent, err := client.CreateAgent("agent0", &ltConfig, &simulcontroller.Config{})
+		require.Error(t, err)
+		require.Contains(t, err.Error(), "could not validate config")
+		require.Nil(t, agent)
+	})
+
+	t.Run("successful creation", func(t *testing.T) {
+		var ltConfig loadtest.Config
+		var ucConfig simulcontroller.Config
+		defaults.Set(&ltConfig)
+		defaults.Set(&ucConfig)
+		agent, err := client.CreateAgent("agent0", &ltConfig, &ucConfig)
+		require.NoError(t, err)
+		require.NotNil(t, agent)
+	})
+
+	t.Run("conflict failure", func(t *testing.T) {
+		var ltConfig loadtest.Config
+		var ucConfig simulcontroller.Config
+		defaults.Set(&ltConfig)
+		defaults.Set(&ucConfig)
+		agent, err := client.CreateAgent("agent0", &ltConfig, &ucConfig)
+		require.Error(t, err)
+		require.Nil(t, agent)
+		require.Contains(t, err.Error(), "load-test agent with id agent0 already exists")
+	})
+}
+
+func TestAgentId(t *testing.T) {
+	// create http.Handler
+	handler := SetupAPIRouter(logger.New(&logger.Settings{}), logger.New(&logger.Settings{}))
+
+	// run server using httptest
+	server := httptest.NewServer(handler)
+	defer server.Close()
+
+	client := NewClient(server.URL, nil)
+	require.NotNil(t, client)
+
+	id := "agent0"
+	agent := createAgent(t, client, id)
+	require.Equal(t, id, agent.Id())
+}
+
+func TestAgentStatus(t *testing.T) {
+	// create http.Handler
+	handler := SetupAPIRouter(logger.New(&logger.Settings{}), logger.New(&logger.Settings{}))
+
+	// run server using httptest
+	server := httptest.NewServer(handler)
+	defer server.Close()
+
+	client := NewClient(server.URL, nil)
+	require.NotNil(t, client)
+
+	id := "agent0"
+	agent := createAgent(t, client, id)
+
+	status, err := agent.Status()
+	require.NoError(t, err)
+	require.Empty(t, status)
+	require.Equal(t, loadtest.Stopped, status.State)
+}
+
+func TestAgentRunStop(t *testing.T) {
```

Comment:
I guess we are not interested in testing for concurrency here? Just checking.

Reply:
If we did that we would be testing the API server more than anything. Each agent/coordinator client instance only shares the http client which is safe for concurrent use.

Reply:
Sure, I am fine with it. Just wanted to confirm.